src/linker/Linker.Dataflow/MethodBodyScanner.cs

```diff
@@ -0,0 +1,825 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+using Mono.Collections.Generic;
+
+namespace Mono.Linker.Dataflow
+{
+	/// <summary>
+	/// Tracks information about the contents of a stack slot
+	/// </summary>
+	class StackSlot
+	{
+		public ValueNode Value { get; set; }
+
+		/// <summary>
+		/// True if the value is on the stack as a byref
+		/// </summary>
+		public bool IsByRef { get; set; }
+
+		public StackSlot ()
+		{
+		}
+
+		public StackSlot (ValueNode value, bool isByRef = false) : this ()
+		{
+			Value = value;
+			IsByRef = isByRef;
+		}
+	}
+
+	abstract partial class MethodBodyScanner
+	{
+		internal ValueNode MethodReturnValue { private set; get; }
+
+		protected virtual void WarnAboutInvalidILInMethod(MethodBody method, int ilOffset)
+		{
+		}
+
+		private void CheckForInvalidStack (Stack<StackSlot> stack, int depthRequired, MethodBody method, int ilOffset)
+		{
+			if (stack.Count < depthRequired) {
+				WarnAboutInvalidILInMethod (method, ilOffset);
+				while (stack.Count < depthRequired)
+					stack.Push (new StackSlot ()); // Push dummy values to avoid crashes.
+												   // Analysis of this method will be incorrect.
+			}
+		}
+
+		private void CheckForInvalidReturnStack (Stack<StackSlot> stack, MethodBody method, int ilOffset)
+		{
+			int numExpectedValuesOnStack = method.Method.ReturnType.MetadataType == MetadataType.Void ? 0 : 1;
+			if (stack.Count != numExpectedValuesOnStack) {
+				WarnAboutInvalidILInMethod (method, ilOffset);
+			}
+		}
+
+		private static void PushUnknown (Stack<StackSlot> stack, int count)
```

Comment:
All usage uses count == 1, why do you need a loop here?

Reply:
Leftover from .NET Native. Removed the count parameter. We can always put it back once it's needed.